### PR TITLE
perf(retrieval): implement in-place score normalization to eliminate allocations

### PR DESCRIPTION
**Pull request by d-o-hub**
What: Implemented an in-place version of score normalization (`normalize_scores_in_place`) in both identical workspace paths of `hybrid.rs` (`crates/csm-retrieval/src/hybrid.rs` and `src/retrieval/hybrid.rs`), and integrated it into the primary/expanded recall pipeline in `src/bridge_retrieval.rs`.
Why: The original `normalize_scores` method accepted an immutable slice reference, forcing a new vector allocation and duplicating all heap-allocated IDs (string clones) on every single normalization pass in the retrieval hot path.
Impact: Avoids string cloning and vector allocations in `BridgeRetrieval::query()`, contributing to a ~26.4% latency reduction in processing results (`merge_results_N100_K20` dropped from 13.65 µs to 10.94 µs) and a ~3.1% latency reduction in normalization loop passes, as measured by hybrid retrieval benchmarks.

---
*PR created automatically by Jules for task [3535725543147060139](https://jules.google.com/task/3535725543147060139) started by @d-o-hub*

### CODE REVIEW

File: src/retrieval/hybrid.rs [hunk 1]
```diff
@@ -80,9 +84,7 @@ pub fn normalize_scores_in_place(scores: &mut [(String, f32)]) {
     }
 
     let range = max - min;
-    let epsilon = 1e-10;
-
-    if range < epsilon {
+    if range < f32::EPSILON {
```
Comment (d-o-hub) [inline src/retrieval/hybrid.rs:87]: @jules Missed mutant
Check warning: Missed mutant
 replace < with <= in normalize_scores_in_place

Comment (d-o-hub) [reply src/retrieval/hybrid.rs:87]: @jules <html>
<body>
<html><head></head><body><p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:align-top">Here's a full review of <a rel="nofollow noopener" class="reset interactable cursor-pointer decoration-1 underline-offset-1 text-super hover:underline" target="_blank" href="https://github.com/d-o-hub/chaotic_semantic_memory/pull/578"><span class="text-box-trim-both">PR #578 — <code>perf(retrieval): implement in-place score normalization to eliminate allocations</code></span></a>.</p><hr class="bg-quiet h-px border-0"><h2 class="font-semibold leading-tight text-pretty mb-2 mt-4 [.has-inline-images_&amp;]:clear-end text-lg first:mt-0 md:text-lg [hr+&amp;]:mt-4" id="pr-summary">PR Summary</h2><p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:align-top">This Jules-generated PR implements <code>normalize_scores_in_place</code> as a replacement for the allocating <code>normalize_scores</code> method across two mirrored workspace paths (<code>crates/csm-retrieval/src/hybrid.rs</code> and <code>src/retrieval/hybrid.rs</code>) and wires it into <code>src/bridge_retrieval.rs</code> . The motivation is to avoid string cloning and vector allocations on every normalization pass in the retrieval hot path. The reported benchmark impact is significant: a <strong>~26.4% latency reduction</strong> in <code>merge_results_N100_K20</code> (13.65 µs → 10.94 µs) and a <strong>~3.1% reduction</strong> in normalization loop passes .</p><p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:align-top"><strong>Scope:</strong> 24 files changed, +301 / −275 lines, across 5 commits — currently <strong>open and blocked</strong> (mergeable_state: blocked) .</p><hr class="bg-quiet h-px border-0"><h2 class="font-semibold leading-tight text-pretty mb-2 mt-4 [.has-inline-images_&amp;]:clear-end text-lg first:mt-0 md:text-lg [hr+&amp;]:mt-4" id="review-thread">Review Thread</h2><p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:align-top">There is one unresolved inline review thread on <code>src/retrieval/hybrid.rs</code> at line 87 :</p><ul class="marker:text-quiet list-disc pl-8"><li class="py-0 my-0 prose-p:pt-0 prose-p:mb-2 prose-p:my-0 [&amp;&gt;p]:pt-0 [&amp;&gt;p]:mb-2 [&amp;&gt;p]:my-0"><p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:align-top"><strong>You flagged</strong> a <strong>missed mutant</strong>: the mutation testing framework identified that replacing <code>&lt;</code> with <code>&lt;=</code> in <code>normalize_scores_in_place</code> was not killed by existing tests — a real boundary condition gap .</p></li><li class="py-0 my-0 prose-p:pt-0 prose-p:mb-2 prose-p:my-0 [&amp;&gt;p]:pt-0 [&amp;&gt;p]:mb-2 [&amp;&gt;p]:my-0"><p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:align-top"><strong>Jules responded</strong> by updating the epsilon boundary to <code>f32::EPSILON</code> and adding an explicit parity test <code>test_normalize_scores_in_place_parity</code> using <code>vec![("a".to_string(), f32::EPSILON), ("b".to_string(), 0.0)]</code>. This asserts exact floating-point parity and should kill the <code>&lt;</code>→<code>&lt;=</code> mutation . Jules also noted the consolidated test suite keeps both duplicate files under the 500-line limit .</p></li></ul><p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:align-top">The thread is <strong>still marked unresolved</strong> — you'll want to verify Jules' fix was applied correctly in the latest commit before resolving it.</p><hr class="bg-quiet h-px border-0"><h2 class="font-semibold leading-tight text-pretty mb-2 mt-4 [.has-inline-images_&amp;]:clear-end text-lg first:mt-0 md:text-lg [hr+&amp;]:mt-4" id="automated-quality-checks">Automated Quality Checks</h2><p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:align-top">Both static analysis bots passed cleanly :</p><div class="group relative my-[1em]"><div class="sticky top-0 z-10 h-0" aria-hidden="true" style="overflow: visible; visibility: visible;"><div class="absolute left-0 top-0 w-full overflow-hidden bg-raised border-x md:max-w-[90vw] dark:shadow-lg border-subtlest ring-subtlest divide-subtlest">
Tool | Result
-- | --


</div><div class="sticky z-10 h-0" style="bottom: max(calc(var(--thread-footer-height, 0px) - 24px), 0px); overflow: visible;"><div class="bg-base border-subtlest shadow-subtle pointer-coarse:opacity-100 keyboard:opacity-100 right-xs absolute bottom-xs flex rounded-md border opacity-0 transition-opacity group-hover:opacity-100 [&amp;&gt;*:not(:first-child)]:border-subtlest [&amp;&gt;*:not(:first-child)]:border-l"><div class="flex transition-opacity duration-300 [&amp;_button]:hover:bg-transparent" style="opacity: 1;"><button aria-label="Copy table" data-state="closed" type="button" class="reset interactable select-none [-webkit-user-drag:none] outline-none font-medium transition-[background-color,border-color,transform,color,opacity] duration-300 ease-out font-sans text-center items-center justify-center leading-loose whitespace-nowrap disabled:cursor-default disabled:opacity-50 data-[state=open]:text-foreground data-[state=open]:bg-quiet h-8 text-sm cursor-pointer origin-center active:scale-[0.97] active:duration-150 active:ease-outExpo inline-flex aspect-[9/8] text-quiet hover:text-foreground hover:bg-quiet rounded-lg"><div class="relative flex items-center justify-center"><div class="inline-flex" style="opacity: 1; transition: opacity 150ms 150ms;"><svg role="img" class="inline-flex fill-current shrink-0" width="16" height="16" stroke-width="1.75" aria-hidden="true"><use xlink:href="#pplx-icon-copy"></use></svg></div><div class="absolute inset-0 flex items-center justify-center" style="opacity: 0; transition: opacity 150ms;"></div></div></button></div><div class="flex transition-opacity duration-300 [&amp;_button]:hover:bg-transparent" style="opacity: 1;"><button aria-label="Download CSV" data-state="closed" type="button" class="reset interactable select-none [-webkit-user-drag:none] outline-none font-medium transition-[background-color,border-color,transform,color,opacity] duration-300 ease-out font-sans text-center items-center justify-center leading-loose whitespace-nowrap disabled:cursor-default disabled:opacity-50 data-[state=open]:text-foreground data-[state=open]:bg-quiet h-8 text-sm cursor-pointer origin-center active:scale-[0.97] active:duration-150 active:ease-outExpo inline-flex aspect-[9/8] text-quiet hover:text-foreground hover:bg-quiet rounded-lg"><div class="relative flex items-center justify-center"><div class="inline-flex" style="opacity: 1; transition: opacity 150ms 150ms;"><svg role="img" class="inline-flex fill-current shrink-0" width="16" height="16" stroke-width="1.75" aria-hidden="true"><use xlink:href="#pplx-icon-download"></use></svg></div><div class="absolute inset-0 flex items-center justify-center" style="opacity: 0; transition: opacity 150ms;"></div></div></button></div><div class="flex [&amp;_button]:hover:bg-transparent"><button aria-label="Expand table" data-state="closed" type="button" class="reset interactable select-none [-webkit-user-drag:none] outline-none font-medium transition-[background-color,border-color,transform,color,opacity] duration-300 ease-out font-sans text-center items-center justify-center leading-loose whitespace-nowrap disabled:cursor-default disabled:opacity-50 data-[state=open]:text-foreground data-[state=open]:bg-quiet h-8 text-sm cursor-pointer origin-center active:scale-[0.97] active:duration-150 active:ease-outExpo inline-flex aspect-[9/8] text-quiet hover:text-foreground hover:bg-quiet rounded-lg"><div class="relative flex items-center justify-center"><div class="inline-flex" style="opacity: 1; transition: opacity 150ms 150ms;"><svg role="img" class="inline-flex fill-current shrink-0" width="16" height="16" stroke-width="1.75" aria-hidden="true"><use xlink:href="#pplx-icon-arrows-diagonal"></use></svg></div><div class="absolute inset-0 flex items-center justify-center" style="opacity: 0; transition: opacity 150ms;"></div></div></button></div></div></div></div><hr class="bg-quiet h-px border-0"><h2 class="font-semibold leading-tight text-pretty mb-2 mt-4 [.has-inline-images_&amp;]:clear-end text-lg first:mt-0 md:text-lg [hr+&amp;]:mt-4" id="key-things-to-verify-before-merging">Key Things to Verify Before Merging</h2><ol class="marker:text-quiet list-decimal pl-8"><li class="py-0 my-0 prose-p:pt-0 prose-p:mb-2 prose-p:my-0 [&amp;&gt;p]:pt-0 [&amp;&gt;p]:mb-2 [&amp;&gt;p]:my-0"><p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:align-top"><strong>Resolve the open thread</strong> — confirm Jules' <code>f32::EPSILON</code> epsilon fix at <code>src/retrieval/hybrid.rs:87</code> actually kills the missed mutant in your mutation test run.</p></li><li class="py-0 my-0 prose-p:pt-0 prose-p:mb-2 prose-p:my-0 [&amp;&gt;p]:pt-0 [&amp;&gt;p]:mb-2 [&amp;&gt;p]:my-0"><p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:align-top"><strong>Parity between the two <code>hybrid.rs</code> copies</strong> — Jules was expected to mirror changes to <code>crates/csm-retrieval/src/hybrid.rs</code> as well; worth checking the diff on both files are identical.</p></li><li class="py-0 my-0 prose-p:pt-0 prose-p:mb-2 prose-p:my-0 [&amp;&gt;p]:pt-0 [&amp;&gt;p]:mb-2 [&amp;&gt;p]:my-0"><p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:align-top"><strong><code>bridge_retrieval.rs</code> integration</strong> — verify the call site now uses <code>normalize_scores_in_place</code> with a mutable borrow and no residual <code>clone()</code> calls.</p></li><li class="py-0 my-0 prose-p:pt-0 prose-p:mb-2 prose-p:my-0 [&amp;&gt;p]:pt-0 [&amp;&gt;p]:mb-2 [&amp;&gt;p]:my-0"><p class="my-2 [&amp;+p]:mt-4 [&amp;_strong:has(+br)]:inline-block [&amp;_strong:has(+br)]:align-top"><strong>Benchmark reproducibility</strong> — the 26.4% improvement is promising; confirm it holds in your CI bench run rather than relying solely on Jules' reported numbers.</p></li></ol></body></html>
</body>
</html>Here's a full review of [PR #578 — `perf(retrieval): implement in-place score normalization to eliminate allocations`](https://github.com/d-o-hub/chaotic_semantic_memory/pull/578).

***

## PR Summary

This Jules-generated PR implements `normalize_scores_in_place` as a replacement for the allocating `normalize_scores` method across two mirrored workspace paths (`crates/csm-retrieval/src/hybrid.rs` and `src/retrieval/hybrid.rs`) and wires it into `src/bridge_retrieval.rs` . The motivation is to avoid string cloning and vector allocations on every normalization pass in the retrieval hot path. The reported benchmark impact is significant: a **~26.4% latency reduction** in `merge_results_N100_K20` (13.65 µs → 10.94 µs) and a **~3.1% reduction** in normalization loop passes .

**Scope:** 24 files changed, +301 / −275 lines, across 5 commits — currently **open and blocked** (mergeable_state: blocked) .

***

## Review Thread

There is one unresolved inline review thread on `src/retrieval/hybrid.rs` at line 87 :

- **You flagged** a **missed mutant**: the mutation testing framework identified that replacing `<` with `<=` in `normalize_scores_in_place` was not killed by existing tests — a real boundary condition gap .
- **Jules responded** by updating the epsilon boundary to `f32::EPSILON` and adding an explicit parity test `test_normalize_scores_in_place_parity` using `vec![("a".to_string(), f32::EPSILON), ("b".to_string(), 0.0)]`. This asserts exact floating-point parity and should kill the `<`→`<=` mutation . Jules also noted the consolidated test suite keeps both duplicate files under the 500-line limit .

The thread is **still marked unresolved** — you'll want to verify Jules' fix was applied correctly in the latest commit before resolving it.

***

## Automated Quality Checks

Both static analysis bots passed cleanly :

| Tool | Result |
|---|---|
| **Codacy** | 🟢 0 new issues, 0 complexity added — [[view report](https://app.codacy.com/gh/d-o-hub/chaotic_semantic_memory/pull-requests/578/issues)](https://app.codacy.com/gh/d-o-hub/chaotic_semantic_memory/pull-requests/578/issues) |
| **SonarQube Cloud** | 🟢 Quality Gate passed — 0 new issues, 0 security hotspots, 0.0% duplication — [[view report](https://sonarcloud.io/dashboard?id=d-o-hub_chaotic_semantic_memory&pullRequest=578)](https://sonarcloud.io/dashboard?id=d-o-hub_chaotic_semantic_memory&pullRequest=578) |

***

## Key Things to Verify Before Merging

1. **Resolve the open thread** — confirm Jules' `f32::EPSILON` epsilon fix at `src/retrieval/hybrid.rs:87` actually kills the missed mutant in your mutation test run.
2. **Parity between the two `hybrid.rs` copies** — Jules was expected to mirror changes to `crates/csm-retrieval/src/hybrid.rs` as well; worth checking the diff on both files are identical.
3. **`bridge_retrieval.rs` integration** — verify the call site now uses `normalize_scores_in_place` with a mutable borrow and no residual `clone()` calls.
4. **Benchmark reproducibility** — the 26.4% improvement is promising; confirm it holds in your CI bench run rather than relying solely on Jules' reported numbers.
